cl_dll/vgui_UnicodeTextImage.cpp

```diff
@@ -0,0 +1,558 @@
+	virtual void GetTextSize(UnicodeTextImage::HFont font, const wchar_t *text, int &wide, int &tall) = 0;
+
+	// notify icons?!?
+	virtual VPANEL GetNotifyPanel() = 0;
+	virtual void SetNotifyIcon(VPANEL context, HTexture icon, VPANEL panelToReceiveMessages, const char *text) = 0;
+
+	// plays a sound
+	virtual void PlaySound(const char *fileName) = 0;
+
+	//!! these functions should not be accessed directly, but only through other vgui items
+	//!! need to move these to seperate interface
+	virtual int GetPopupCount() = 0;
+	virtual VPANEL GetPopup(int index) = 0;
+	virtual bool ShouldPaintChildPanel(VPANEL childPanel) = 0;
+	virtual bool RecreateContext(VPANEL panel) = 0;
+	virtual void AddPanel(VPANEL panel) = 0;
+	virtual void ReleasePanel(VPANEL panel) = 0;
+	virtual void MovePopupToFront(VPANEL panel) = 0;
+	virtual void MovePopupToBack(VPANEL panel) = 0;
+
+	virtual void SolveTraverse(VPANEL panel, bool forceApplySchemeSettings = false) = 0;
+	virtual void PaintTraverse(VPANEL panel) = 0;
+
+	virtual void EnableMouseCapture(VPANEL panel, bool state) = 0;
+
+	// returns the size of the workspace
+	virtual void GetWorkspaceBounds(int &x, int &y, int &wide, int &tall) = 0;
+
+	// gets the absolute coordinates of the screen (in windows space)
+	virtual void GetAbsoluteWindowBounds(int &x, int &y, int &wide, int &tall) = 0;
+
+	// gets the base resolution used in proportional mode
+	virtual void GetProportionalBase(int &width, int &height) = 0;
+
+	virtual void CalculateMouseVisible() = 0;
+	virtual bool NeedKBInput() = 0;
+
+	virtual bool HasCursorPosFunctions() = 0;
+	virtual void SurfaceGetCursorPos(int &x, int &y) = 0;
+	virtual void SurfaceSetCursorPos(int x, int y) = 0;
+
+
+	// SRC only functions!!!
+	virtual void DrawTexturedPolygon(VGuiVertex *pVertices, int n) = 0;
+	virtual int GetFontAscent(UnicodeTextImage::HFont font, wchar_t wch) = 0;
+
+	// web browser
+	virtual void SetAllowHTMLJavaScript(bool state) = 0;
+
+	virtual void SetLanguage(const char *pchLang) = 0;
+
+	virtual const char *GetLanguage() = 0;
+
+	virtual bool DeleteTextureByID(int id) = 0;
+
+	virtual void DrawUpdateRegionTextureBGRA(int nTextureID, int x, int y, const unsigned char *pchData, int wide, int tall) = 0;
+
+	virtual void DrawSetTextureBGRA(int id, const unsigned char *pchData, int wide, int tall) = 0;
+
+	virtual void CreateBrowser(VPANEL panel, IHTMLResponses *pBrowser, bool bPopupWindow, const char *pchUserAgentIdentifier) = 0;
+
+	virtual void RemoveBrowser(VPANEL panel, IHTMLResponses *pBrowser) = 0;
+
+	virtual IHTMLChromeController *AccessChromeHTMLController() = 0;
+};
+
+#define VGUI_SURFACE_INTERFACE_VERSION "VGUI_Surface026"
+
+class ILocalize : public IBaseInterface
+{
+public:
+	virtual void Something1() = 0;
+	virtual void Something2() = 0;
+	virtual void Something3() = 0;
+
+	// Converts UTF-8 to WString
+	virtual int ConvertANSIToUnicode(const char *ansi, wchar_t *unicode, int unicodeBufferSizeInBytes) = 0;
```

Comment:
What happens if there's a 3+-byte character like an emoji and `wchar_t` is 16 bytes?

Reply:
VGUI2 only supports Basic Multilingual Plane (0000-FFFF). Such characters cannot be typed in the console or settings dialog.